using method name to find unqualified Stacktraces in Java Stack Trace

SougandhS opened the pull request:
Using method name to find unqualified Stacktraces
Fix for - https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/115

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See https://github.com/eclipse-jdt/.github/security/policy
-->

## What it does
<!-- Include relevant issues and describe how they are addressed. -->
Resolves the ambiguity in results generated when fully qualified class name is not provided in java stack trace console
## How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

SougandhS commented:
Hi @jukzi 
Made changes for enhancement you proposed -> [115](https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/115) .
Could you please review this PR ?

jukzi commented on org.eclipse.jdt.debug.tests/console tests/org/eclipse/jdt/debug/tests/console/JavaStackTraceAmbiguityTest.java:
```diff
@@ -0,0 +1,73 @@
+package org.eclipse.jdt.debug.tests.console;
```
every .java file needs a copyright header. please add.

SougandhS replied:
Understood jukzi. will add on every .java files

jukzi commented on org.eclipse.jdt.debug.tests/testfiles/Ambiguity/a/Sample.java:
```diff
@@ -0,0 +1,37 @@
+/*******************************************************************************
+ * Copyright (c) 2024 IBM Corporation and others.
```
when adding new files you may change it to your or employers name

SougandhS replied:
I can go with my employers name.

jukzi commented on org.eclipse.jdt.debug.tests/testfiles/Ambiguity/a/Sample.java:
```diff
@@ -0,0 +1,37 @@
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
```
company

SougandhS replied:
Since I am part of IBM.

jukzi commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/console/JavaStackTraceHyperlink.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2000, 2017 IBM Corporation and others.
+ * Copyright (c) 2024 IBM Corporation and others.
```
the initial date must not be changed. the second date may be updated, but does not need to

SougandhS replied:
okay, will update it like that

jukzi commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/console/JavaStackTraceHyperlink.java:
```diff
@@ -190,8 +195,37 @@ public IStatus runInUIThread(IProgressMonitor monitor) {
+								IMethod[] methods = source.getMethods();
+								for (IMethod method : methods) {
+									int indexOfClosing = method.toString().indexOf(")"); //$NON-NLS-1$
+									int indexOfStart = method.toString().substring(0,indexOfClosing+1).lastIndexOf(" "); //$NON-NLS-1$
```
`lastIndexOf(" ")` may work in this situation but generally you should avoid matching for space, as other or multiple whitespaces would be legal too. So search for something meaningfull

SougandhS replied:
In this specific case multiple whitespaces are occurred prior to the actual function name, here the toString() always returns values in this format -> _"**static void test() [in Sample [in Sample.java [in <default> [in src [in New]]]]]**"_
and for every extracted methods there was only one closing and one opening, which is part of the actual function name, so my approach was put the closing brace as the limit of the substring and access the last whitespace which always results in getting the whitespace prior to the actual function name, there weren't any direct methods for extracting method name & signature only

jukzi commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/console/JavaStackTraceHyperlink.java:
```diff
@@ -190,8 +195,37 @@ public IStatus runInUIThread(IProgressMonitor monitor) {
 					List<Object> matches = (List<Object>) source;
+					List<Object> exactMatchesFiltered = new ArrayList<>();
+					String originalHyperLink2 = originalHyperLink;
+					int firstMethodStartIndex = originalHyperLink2.indexOf("."); //$NON-NLS-1$
```
to avoid `//$NON-NLS-1$` it's often possible to use single characters like `indexOf('.')`

SougandhS replied:
Okay

jukzi reviewed (changes requested):
i tested the change with
`Project.writeDescription(int) line: 1402` in platform workspace and ended up in exception:
```
java.lang.ClassCastException: class org.eclipse.jdt.internal.core.BinaryType cannot be cast to class org.eclipse.jdt.internal.core.SourceType (org.eclipse.jdt.internal.core.BinaryType and org.eclipse.jdt.internal.core.SourceType are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @35f7b857)
	at org.eclipse.jdt.internal.debug.ui.console.JavaStackTraceHyperlink$3.runInUIThread(JavaStackTraceHyperlink.java:215)
	at org.eclipse.ui.progress.UIJob.lambda$0(UIJob.java:148)
```

jukzi commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/console/JavaStackTraceHyperlink.java:
```diff
@@ -183,15 +196,45 @@ public IStatus runInUIThread(IProgressMonitor monitor) {
+						String firstMethod = originalHyperLink2.substring(firstMethodStartIndex + 1, firstMethodClosing + 1);
+						for (Object obj : matches) {
+							@SuppressWarnings("restriction")
+							SourceType source = (SourceType) obj;
```
use `IType `and do an instanceof check to prevent errors

SougandhS replied:
Okay, will do this change right now

SougandhS replied:
Changes done and tested now.

jukzi reviewed (changes requested):
no exceptions, but feels like it still has issues to find methods with two parameters. here are some examples:
`Realm.runWithDefault(Realm, Runnable) line: 339`
`PlatformUI.createAndRunWorkbench(Display, WorkbenchAdvisor) line: 173	`
![image](https://github.com/user-attachments/assets/0f8bd981-5077-4cce-a20c-6beec0f0ed95)
should have found org.eclipse.core.databinding.observable.Realm
![image](https://github.com/user-attachments/assets/c7ab81f2-4e09-453c-a877-c4d6ecd29ed7)
`DirectMethodHandleAccessor.invokeImpl(Object, Object[]) line: 155`
![image](https://github.com/user-attachments/assets/27d2a35d-9a73-45f7-bd70-2c9c20ad065d)

`Method.invoke(Object, Object...) line: 580`
![image](https://github.com/user-attachments/assets/3668f0cf-8355-4779-9097-39f9af02b184)
`Main.invokeFramework(String[], URL[]) line: 668	`

jukzi commented:
bonus: Nested Class (with $) are also not found:
`AbstractQueuedSynchronizer$ConditionNode.block() line: 519	`
![image](https://github.com/user-attachments/assets/46d29de6-0e69-4ee7-8e9b-cc32cb49dd52)

SougandhS commented:
@jukzi thanks for the additional cases, will make changes and fix the issues.

jukzi commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/console/JavaStackTraceHyperlink.java:
```diff
@@ -183,18 +199,180 @@ public IStatus runInUIThread(IProgressMonitor monitor) {
+					}
+					if (exactMatchesFiltered.size() == 2) {
+						try { // Occurred only in child eclipse
+							if (exactMatchesFiltered.get(0) instanceof BinaryType && exactMatchesFiltered.get(1) instanceof BinaryType) {
```
can you get rid of the compiler warnings here? `type 'BinaryType' is not API`
Also its handsome to use pattern matching instanceof where possible:
`(x instanceof IType y) y.method()`

SougandhS replied:
Sure Jukzi, will do like you suggested.

jukzi commented:
Please figure out why the Build fails InstanceMainMethodsTests - may be related to the Java23 merge @jarthana

jukzi commented:
The tests are also failing on master:
https://ci.eclipse.org/jdt/job/eclipse.jdt.debug-github/job/master/119/
If failing tests are unrelated please create an Issue for it and then we could ignore them for the current PR

SougandhS commented:
> The tests are also failing on master: https://ci.eclipse.org/jdt/job/eclipse.jdt.debug-github/job/master/119/ If failing tests are unrelated please create an Issue for it and then we could ignore them for the current PR

these tests are unrelated to the current changes, will raise issue for those now

SougandhS commented:
@jukzi those 3 tests are now passing for all builds now. 
shall we close Issue #506 ?

jukzi reviewed (approved):
I have tested and it works like a charm. Very good improvement. thanks.
Found two other stacktraces that do not work yet(wrong linenumber jumped to):
```
ScheduledThreadPoolExecutor(ThreadPoolExecutor).runWorker(ThreadPoolExecutor$Worker) line: 1130	
ThreadPoolExecutor.runWorker(ThreadPoolExecutor$Worker) line: 1130
```
![image](https://github.com/user-attachments/assets/24751396-a3b2-4eee-914c-bee0f1d29a2a)

Please let me know if you want to fix those too in this PR. Otherwise we can also submit  the current state as it already is an improvement as is.

SougandhS commented:
> I have tested and it works like a charm. Very good improvement. thanks. Found two other stacktraces that do not work yet(wrong linenumber jumped to):
> 
> ```
> ScheduledThreadPoolExecutor(ThreadPoolExecutor).runWorker(ThreadPoolExecutor$Worker) line: 1130	
> ThreadPoolExecutor.runWorker(ThreadPoolExecutor$Worker) line: 1130
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/51790620/369797609-24751396-a3b2-4eee-914c-bee0f1d29a2a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcwOTkzNzYsIm5iZiI6MTcyNzA5OTA3NiwicGF0aCI6Ii81MTc5MDYyMC8zNjk3OTc2MDktMjQ3NTEzOTYtYTNiMi00ZWVlLTkxNGMtYmVlMGYxZDI5YTJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTIzVDEzNDQzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc1MzcxZjc1MTg4OWMwYTJjNDFmMzY1NWE5YTNkNGUwNmVhZmRkNzEwOTQ3NWNjNDMzNjVjMzBmZGFkNjM0ZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RQ2klw208M6KOY0oDBJoImXbwxhzXGMjuB9hG26i-JE)
> 
> Please let me know if you want to fix those too in this PR. Otherwise we can also submit the current state as it already is an improvement as is.

hi @jukzi ,fixed those and added additional tests for that in the latest commit.

SougandhS commented:
@jukzi all commits have squashed into one now. (includes additional case's fix)

jukzi commented:
@SougandhS thanks! let's try it in real :-)

SougandhS commented:
Thank you @jukzi :)
